File: Queue.cpp
// Using DoulbeLinkedList to implement a Queue
#include <iostream>
using namespace std;

// forward declaration of class Queue
template<typename NODETYPE> class Queue; 

template<typename NODETYPE>
class Node
{
    friend class Queue<NODETYPE>; 
    /*
    two reasons:
    1. let class Queue being able to specialize its member funcitons based on the template 
    2. allow class Queue access the priviate members of the class Node
    */
private:
    NODETYPE item;
    Node<NODETYPE>* prev;
    Node<NODETYPE>* next;

public:
    explicit Node(const NODETYPE& item)
        : item(item), prev(nullptr), next(nullptr) 
    {}
};

template<typename NODETYPE>
class Queue
{
private:
    size_t q_size;
    Node<NODETYPE>* head;
    Node<NODETYPE>* tail;

public:
    Queue(): head(nullptr), tail(nullptr), q_size(0) {}
    
    ~Queue()
    {
        if (!empty())
        {
            Node<NODETYPE>* curr = head;
            Node<NODETYPE>* temp;
            while (curr != nullptr)
            {
                temp = curr;
                curr = curr->next;
                delete temp;
            }            
        }
    }

    bool empty() const { return head == nullptr; }

    NODETYPE top() const { return head->item; }

    size_t size() const { return q_size; }

    void enqueue(const NODETYPE& item)
    {
        Node<NODETYPE>* pnode = new Node<NODETYPE>(item);
        if (empty())
            head = tail = pnode;
        else
        {
            tail->next = pnode; // link the current last node to the new last node
            pnode->prev = tail; // link the new last node back to the current last node
            tail = pnode; // update tail pointer
        }
        q_size++;
    }

    NODETYPE dequeue()
    {
        NODETYPE item;
        if (!empty())
        {
            item = head->item;
            if (head == tail) // only one element in the queue
                head = tail = nullptr;
            else
            {
                Node<NODETYPE>* temp = head; // hold node to delete
                head = head->next; // update head pointer
                head->prev = nullptr; // break 
                delete temp;
            }
            q_size--;
        }
        return item;
    }

    void print() const
    {
        if (empty())
            cout << "Queue is empty\n\n";
        else
        {
            Node<NODETYPE>* curr = head;
            cout << "{ ";
            while (curr)
            {
                cout << curr->item << " ";
                curr = curr->next;
            }
            cout << "}\n\n";
        }
    }
};


int main()
{
    Queue<int> q;
    cout << "q.empty()? " << q.empty() << endl;
    for (int i = 0; i < 4; i++)
        q.enqueue(i);
    cout << "q.empty()? " << q.empty() << endl;
    cout << "q.size() = " << q.size() << endl;
    cout << "q.top() = " << q.top() << endl;
    cout << "q = "; q.print();

    auto first_item = q.dequeue();
    cout << "first_item = " << first_item << endl;
    cout << "q.empty()? " << q.empty() << endl;
    cout << "q.size() = " << q.size() << endl;
    cout << "q.top() = " << q.top() << endl;    
    cout << "q = "; q.print();
}

File: Queue.py
# implemented by doulbe linked list
# O(1) dequeue/enqueue

class DoubleLinkedNode:
    def __init__(self, item):
        self.item = item
        self.prev = None
        self.next = None

class Queue:
    def __init__(self):
        self._size = 0
        self.head = None
        self.tail = None
    
    def enqueue(self, item): # move tial
        node = DoubleLinkedNode(item)
        if self.tail is None: # no item in the queue
            self.tail = node
            self.head = self.tail
        else: # already have items in the queue
           self.tail.next = node # link the last item to the new node
           node.prev = self.tail # link the new node to the last node
           self.tail = self.tail.next # move the tail pointer to the current last node
        self._size += 1
    
    def dequeue(self):
        if self.head is None: # no item in the queue
            return None
        item = self.head.item # get the return item
        self.head = self.head.next # move the head pointer to the current head
        self.head.prev = None # link the head should not have previous node
        self._size -= 1
        return item 
    
    def size(self):
        return self._size

    def top(self):
        return self.head.item

    def empty(self):
        return self._size == 0
    

if __name__ == "__main__":
    q = arrayQueue()
    print(f"q.empty() = {q.empty()}")
    for i in range(4):
        q.enqueue(i)
    print(f"q.empty() = {q.empty()}")
    print(f"q.size() = {q.size()}")    
    print(f"q.top() = {q.top()}")

    first_item = q.dequeue()
    print(f"first_item = {first_item}")
    print(f"q.empty() = {q.empty()}")
    print(f"q.size() = {q.size()}")    
    print(f"q.top() = {q.top()}")
    